examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+  // respectively. The last argument "alpha" is used while executing line search: it acts as the tentative weight of
+  // Newton's update term.
+
+  template <int dim>
+  void Navier_Stokes_Newton<dim>::assemble_NavierStokes_system(bool initial_step,
+      bool assemble_matrix,
+      bool assemble_rhs)
+  {
+    if (assemble_matrix)
+      {
+        system_matrix    = 0;
+      }
+
+    if (assemble_rhs)
+      {
+        system_rhs       = 0;
+      }
+
+
+
+    QGauss<dim>   quadrature_formula(degree+2);
+
+    FEValues<dim> fe_values (fe,
+                             quadrature_formula,
+                             update_values |
+                             update_quadrature_points |
+                             update_JxW_values |
+                             update_gradients );
+
+    const unsigned int   dofs_per_cell = fe.dofs_per_cell;
+    const unsigned int   n_q_points    = quadrature_formula.size();
+
+    const FEValuesExtractors::Vector velocities (0);
+    const FEValuesExtractors::Scalar pressure (dim);
+
+    FullMatrix<double>   local_matrix (dofs_per_cell, dofs_per_cell);
+    Vector<double>       local_rhs    (dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    // For the linearized system, we create temporary storages for present velocity and its divergence and gradient, and
+    // present pressure. In practice, they are all obtained through their sharp functions at quadrature points. This work
+    // can be done by FEValues::get_function_values, FEValues::get_function_gradients and FEValues::get_function_divergences.
+
+    std::vector<Tensor<1, dim>>   present_velocity_values    (n_q_points);
+    std::vector<Tensor<2, dim>>   present_velocity_gradients (n_q_points);
+    std::vector<double>           present_velocity_divergence(n_q_points);
+    std::vector<double>           present_pressure_values    (n_q_points);
+
+    std::vector<double>           div_phi_u                 (dofs_per_cell);
+    std::vector<Tensor<1, dim>>   phi_u                     (dofs_per_cell);
+    std::vector<Tensor<2, dim>>   grad_phi_u                (dofs_per_cell);
+    std::vector<double>           phi_p                     (dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+
+    for (; cell!=endc; ++cell)
+      {
+        fe_values.reinit(cell);
+
+        local_matrix = 0;
+        local_rhs    = 0;
+
+
+        fe_values[velocities].get_function_values(evaluation_point,
+                                                  present_velocity_values);
+
+        fe_values[velocities].get_function_gradients(evaluation_point,
+                                                     present_velocity_gradients);
+
+        fe_values[velocities].get_function_divergences(evaluation_point,
+                                                       present_velocity_divergence);
```

Comment:
I know it does not really matter performance-wise but I'd still avoid this call and rather compute the divergence as `trace(present_velocity_gradients[q])` in the rhs computations below. Why recompute things that are already almost there?